pkg/election/election.go

```diff
@@ -0,0 +1,270 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package election
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math/rand"
+	"os"
+	"path"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/go-zookeeper/zk"
+	logrus "github.com/sirupsen/logrus"
+)
+
+var (
+	errNodeWatchClosed = errors.New("watch channel on znode was closed")
+
+	defaultMaxRandomWaitDuration = 2 * time.Second
+)
+
+type ZooKeeperElection struct {
+	candidateID           string
+	log                   *logrus.Entry
+	basePath              string
+	proposalNodePath      string
+	sessionTimeout        time.Duration
+	maxRandomWaitDuration time.Duration
+	zk                    zkClient
+}
+
+func NewZooKeeperElection(servers []string, electionPath string, candidateID string, sessionTimeout time.Duration, debug bool, l *logrus.Entry) (*ZooKeeperElection, error) {
+	zlog := l.WithField("name", "election-zk-client")
+
+	zk, err := connectZkClient(servers, sessionTimeout, debug, zlog)
+	if err != nil {
+		return nil, fmt.Errorf("zookeeper connection: %w", err)
+	}
+
+	return newElectionWithZkClient(zk, electionPath, candidateID, sessionTimeout, zlog), nil
+}
+
+func newElectionWithZkClient(zk zkClient, electionPath string, candidateID string, sessionTimeout time.Duration, l *logrus.Entry) *ZooKeeperElection {
+	return &ZooKeeperElection{
+		zk:                    zk,
+		basePath:              electionPath,
+		candidateID:           candidateID,
+		sessionTimeout:        sessionTimeout,
+		maxRandomWaitDuration: defaultMaxRandomWaitDuration,
+		log:                   l,
+	}
+}
+
+func (e *ZooKeeperElection) createProposalNode() error {
+	hostname, err := os.Hostname()
+	if err != nil {
+		hostname = "(unknown)"
+	}
+
+	candidateInfo := map[string]string{
+		"candidateId":  e.candidateID,
+		"hostname":     hostname,
+		"pid":          fmt.Sprint(os.Getpid()),
+		"proposalTime": time.Now().String(),
+		// TODO add session timeout
+	}
+
+	data, err := json.Marshal(candidateInfo)
+	if err != nil {
+		data = []byte(fmt.Sprintf(`{"candidateId": "%s"}`, e.candidateID))
+	}
+
+	ownPath, err := e.zk.CreateNodeSequenceEphemeral(e.basePath+"/proposal-", data)
+	if err != nil {
+		return fmt.Errorf("zk create proposal znode: %w", err)
+	}
+
+	e.log.Debugf("own proposal zk node %s", ownPath)
+	e.proposalNodePath = ownPath
+
+	return nil
+}
+
+func (e *ZooKeeperElection) deleteProposalNode() error {
+	err := e.zk.Delete(e.proposalNodePath)
+	if err != nil {
+		return fmt.Errorf("delete own proposal node: %s: %w", e.proposalNodePath, err)
+	}
+
+	e.log.Debugf("successfully deleted own proposal node %s", e.proposalNodePath)
+
+	return nil
+}
+
+func (e *ZooKeeperElection) createElectionPath() error {
+	currentPath := "/"
+
+	for _, part := range strings.Split(e.basePath, "/") {
+		currentPath = path.Join(currentPath, part)
```

Comment:
Not sure it matters, but I'm thinking the "path" module must be designed to handle filesystem paths, not zookeeper paths, hence it might not be the best tool to build zookeeper paths (think Windows platforms :stuck_out_tongue_closed_eyes: )

Reply:
In Go the path module is not really a separator abstraction (unlike other ecosystems), it always uses `/`. I think the `path/filepath` module does this instead. I think `path` is a good tool here because it deals nicely with empty parts.